File: ExtremeStudio/ExtremeParser/Extras/FunctionParameters.vb
Public Class FunctionParameters

    Public ReadOnly Property paramsText
        Get
            Return paramText
        End Get
    End Property

    Public Integers As New Dictionary(Of String, String)
    Public Arrays As New Dictionary(Of String, String)
    Public Floats As New Dictionary(Of String, String)
    Public pawnDoc As PawnDoc

    Private OthersVar As New Dictionary(Of String, String) 'A function is created to handle this.
    Private paramText As String

    Public Enum varTypes
        TYPE_FLOAT
        TYPE_INTEGER
        TYPE_ARRAY
        TYPE_OTHER
        TYPE_TAGGED
    End Enum
    Public Shared Function getVarType(var As String) As varTypes
        If var.StartsWith("Float:") Then
            Return varTypes.TYPE_FLOAT
        ElseIf var.EndsWith("]") And var.Contains("[") Then
            Return varTypes.TYPE_ARRAY
        ElseIf var.Contains(":") Then
            Return varTypes.TYPE_TAGGED
        Else
            Return varTypes.TYPE_INTEGER
        End If
    End Function

    Public Sub New(params As String)
        Dim trimmedParams = params.Replace(" ", "")
        paramText = trimmedParams
        Dim prms As String() = trimmedParams.Split(",")
        For Each Str As String In prms
            If Str = "" Or Str = Nothing Then Continue For

            Dim def As String = ""
            If Str.Contains("=") Then def = Str.Substring(Str.IndexOf("=")).Trim()

            If getVarType(Str) = varTypes.TYPE_FLOAT Then
                Str = Str.Remove(0, 6)
                Floats.Add(Str, def)
            ElseIf getVarType(Str) = varTypes.TYPE_ARRAY Then

                Dim done As Boolean = False
                While done = False
                    If Str.Contains("[") And Str.Contains("]") Then
                        Str = Str.Remove(Str.IndexOf("["), (Str.IndexOf("]") - Str.IndexOf("[")) + 1)
                    Else
                        done = True
                    End If
                End While

                Arrays.Add(Str, def)
            ElseIf getVarType(Str) = varTypes.TYPE_TAGGED Then
                OthersVar.Add(Str, def)
            Else 'Its an integer obv..
                Integers.Add(Str, def)
            End If
        Next
    End Sub

    Public Function Others(tag As String) As List(Of String)
        Dim lst As New List(Of String)
        For Each Str As String In OthersVar.Keys
            Dim vars As String() = OthersVar(Str).Split(":", 2, StringSplitOptions.None)
            If vars(0) = tag Then
                lst.Add(vars(1))
            End If
        Next
        Return lst
    End Function

    Public Enum returnType
        AS_ENUM
        AS_STRING
    End Enum
    Public Function GetParameterType(paramName As String, returnType As returnType)
        For Each Str As String In Integers.Keys
            If Integers(Str) = paramName Then
                If returnType = FunctionParameters.returnType.AS_STRING Then
                    Return "Integer"
                ElseIf returnType = FunctionParameters.returnType.AS_ENUM Then
                    Return varTypes.TYPE_INTEGER
                End If
            End If
        Next
        For Each Str As String In Arrays.Keys
            If Arrays(Str) = paramName Then
                If returnType = FunctionParameters.returnType.AS_STRING Then
                    Return "Array"
                ElseIf returnType = FunctionParameters.returnType.AS_ENUM Then
                    Return varTypes.TYPE_ARRAY 
                End If
            End If
        Next
        For Each Str As String In Floats.Keys
            If Floats(Str) = paramName Then
                If returnType = FunctionParameters.returnType.AS_STRING Then
                    Return "Float"
                ElseIf returnType = FunctionParameters.returnType.AS_ENUM Then
                    Return varTypes.TYPE_FLOAT
                End If
            End If
        Next

        If returnType = FunctionParameters.returnType.AS_STRING Then
            Return "Other"
        ElseIf returnType = FunctionParameters.returnType.AS_ENUM Then
            Return varTypes.TYPE_OTHER
        End If
        Return Nothing
    End Function
End Class

File: ExtremeStudio.Parser/Global/FunctionParameters.cs
// VBConversions Note: VB project level imports
using System.Collections.Generic;
using System;
using System.Threading.Tasks;
using System.Diagnostics;
using System.Data;
using System.Xml.Linq;
using Microsoft.VisualBasic;
using System.Collections;
using System.Linq;
using ExtremeStudio.Parser.Parsers;

// End of VB project level imports


namespace ExtremeStudio.Parser.Global
{
	public class FunctionParameters
	{
		
        public string ParamsText
		{
			get
			{
				return _paramText;
			}
		}
		
		public Dictionary<string, string> Integers = new Dictionary<string, string>();
		public Dictionary<string, string> Arrays = new Dictionary<string, string>();
		public Dictionary<string, string> Floats = new Dictionary<string, string>();
		public PawnDocParser PawnDoc;
		
		private Dictionary<string, string> _othersVar = new Dictionary<string, string>(); //A function is created to handle this.
		private string _paramText;
		
		public enum VarTypes
		{
			TypeFloat,
			TypeInteger,
			TypeArray,
			TypeOther,
			TypeTagged
		}
		public static VarTypes GetVarType(string var)
		{
			if (var.StartsWith("Float:"))
			{
				return VarTypes.TypeFloat;
			}
			else if (var.EndsWith("]") && var.Contains("["))
			{
				return VarTypes.TypeArray;
			}
			else if (var.Contains(":"))
			{
				return VarTypes.TypeTagged;
			}
			else
			{
				return VarTypes.TypeInteger;
			}
		}
		
		public FunctionParameters(string @params)
		{
			var trimmedParams = @params.Replace(" ", "");
			_paramText = trimmedParams;
			string[] prms = trimmedParams.Split(",".ToCharArray());
			foreach (string str in prms)
			{
				if (string.IsNullOrEmpty(str)|| string.IsNullOrEmpty(str))
				{
					continue;
				}
				
				string def = "";
				if (str.Contains("="))
				{
					def = System.Convert.ToString(str.Substring(str.IndexOf("=")).Trim());
				}
				
				if (GetVarType(str) == VarTypes.TypeFloat)
				{
					// str = str.Remove(0, 6); VBConversions Warning: A foreach variable can't be assigned to in C#.
					Floats.Add(str, def);
				}
				else if (GetVarType(str) == VarTypes.TypeArray)
				{
					
					bool done = false;
					while (done == false)
					{
						if (str.Contains("[") && str.Contains("]"))
						{
							// str = str.Remove(str.IndexOf("["), (str.IndexOf("]") - str.IndexOf("[")) + 1); VBConversions Warning: A foreach variable can't be assigned to in C#.
						}
						else
						{
							done = true;
						}
					}
					
					Arrays.Add(str, def);
				}
				else if (GetVarType(str) == VarTypes.TypeTagged)
				{
					_othersVar.Add(str, def);
				}
				else //Its an integer obv..
				{
					Integers.Add(str, def);
				}
			}
		}
		
		public List<string> Others(string tag)
		{
			return ((from str in _othersVar.Keys let vars = _othersVar[str].Split(':', '2', System.Convert.ToChar(StringSplitOptions.None)) select vars).Where(vars => vars[0] == tag).Select(vars => vars[1])).ToList();
		}
		
		public enum ReturnType
		{
			AsEnum,
			AsString
		}
		public dynamic GetParameterType(string paramName, ReturnType returnType)
		{
			VarTypes ret = GetVarType(paramName);
			
			if (returnType == ReturnType.AsEnum)
			{
				return ret;
			}
			else
			{
				if (ret == VarTypes.TypeArray)
				{
					return "Array";
				}
				else if (ret == VarTypes.TypeFloat)
				{
					return "Float";
				}
				else if (ret == VarTypes.TypeInteger)
				{
					return "Integer";
				}
				else if (ret == VarTypes.TypeOther)
				{
					return "Other";
				}
				else if (ret == VarTypes.TypeTagged)
				{
					return "Tagged";
				}
			}
			return null;
		}
	}
	
}
